[Merged by Bors] - Optimize validator duties

paulhauner opened the pull request:
## Issue Addressed

Closes #2052

## Proposed Changes

- Refactor the attester/proposer duties endpoints in the BN
    - Performance improvements
    - Fixes some potential inconsistencies with the dependent root fields.
    - Removes `http_api::beacon_proposer_cache` and just uses the one on the `BeaconChain` instead.
    - Move the code for the proposer/attester duties endpoints into separate files, for readability.
- Refactor the `DutiesService` in the VC
    - Required to reduce the delay on broadcasting new blocks.
    - Gets rid of the `ValidatorDuty` shim struct that came about when we adopted the standard API.
    - Separate block/attestation duty tasks so that they don't block each other when one is slow.
- In the VC, use `PublicKeyBytes` to represent validators instead of `PublicKey`. `PublicKey` is a legit crypto object whilst `PublicKeyBytes` is just a byte-array, it's much faster to clone/hash `PublicKeyBytes` and this change has had a significant impact on runtimes.
    - Unfortunately this has created lots of dust changes.
 - In the BN, store `PublicKeyBytes` in the `beacon_proposer_cache` and allow access to them. The HTTP API always sends `PublicKeyBytes` over the wire and the conversion from `PublicKey` -> `PublickeyBytes` is non-trivial, especially when queries have 100s/1000s of validators (like Pyrmont).
 - Add the `state_processing::state_advance` mod which dedups a lot of the "apply `n` skip slots to the state" code.
    - This also fixes a bug with some functions which were failing to include a state root as per [this comment](https://github.com/sigp/lighthouse/blob/072695284f7eff82c51f79bc921ad942fea7483a/consensus/state_processing/src/state_advance.rs#L69-L74). I couldn't find any instance of this bug that resulted in anything more severe than keying a shuffling cache by the wrong block root.
 - Swap the VC block service to use `mpsc` from `tokio` instead of `futures`. This is consistent with the rest of the code base.
    
~~This PR *reduces* the size of the codebase :tada:~~ It *used* to reduce the size of the code base before I added more comments. 

## Observations on Prymont

- Proposer duties times down from peaks of 450ms to consistent <1ms.
- Current epoch attester duties times down from >1s peaks to a consistent 20-30ms.
- Block production down from +600ms to 100-200ms.

## Additional Info

- ~~Blocked on #2241~~
- ~~Blocked on #2234~~

## TODO

- [x] ~~Refactor this into some smaller PRs?~~ Leaving this as-is for now.
- [x] Address `per_slot_processing` roots.
- [x] Investigate slow next epoch times. Not getting added to cache on block processing?
- [x] Consider [this](https://github.com/sigp/lighthouse/blob/072695284f7eff82c51f79bc921ad942fea7483a/beacon_node/store/src/hot_cold_store.rs#L811-L812) in the scenario of replacing the state roots

pawanjay176 committed:
Attestation service uses sleep instead of interval

pawanjay176 committed:
Duties service uses sleep

pawanjay176 committed:
Fork service uses sleep

pawanjay176 committed:
Create a new task for every duties update

pawanjay176 committed:
Merge branch 'unstable' into vc-time

pawanjay176 committed:
lint

pawanjay176 committed:
Merge branch 'unstable' into vc-time

pawanjay176 committed:
Merge branch 'unstable' into vc-time

pawanjay176 committed:
Cleanup

pawanjay176 committed:
Handle error case

pawanjay176 committed:
Notifier uses sleep

pawanjay176 committed:
Merge branch 'unstable' into vc-time

michaelsproul committed:
Use the database to persist the pubkey cache

paulhauner committed:
Add first pass at using pre-state for block prod

paulhauner committed:
Add warnings for late block broadcasts

paulhauner committed:
Move warnings for late blocks

paulhauner committed:
Add comments

paulhauner committed:
Use crit for late blocks

paulhauner committed:
Tidy docs, add sanity check

paulhauner committed:
Add more metrics for duties

paulhauner committed:
Refactor proposer duties endpoint

paulhauner committed:
Fix bug in state advance timer

paulhauner committed:
Remove old beacon proposer cache

paulhauner committed:
First refactor of attester duties

paulhauner committed:
Merge branch 'pr/2234' into faster-duties

paulhauner committed:
Add pubkey bytes struct

paulhauner committed:
Almost complete duties refactor

paulhauner committed:
Add almost-compiling refactor

paulhauner committed:
Remove old duties service code

paulhauner committed:
Merge branch 'unstable' into faster-duties

paulhauner committed:
Fix compile errors

paulhauner committed:
Remove old validator_duty struct

paulhauner committed:
Fix attestation, indices bugs

paulhauner committed:
Fix clippy lints, add metrics

paulhauner committed:
Specify integer types

paulhauner committed:
Revert "Specify integer types"

This reverts commit 60e1a7dbc73db1b9b8870b1e449ab0c1cc39b881.

paulhauner committed:
Only subscribe for later slots

paulhauner committed:
Specify integers as slots

paulhauner committed:
Use saturating math

paulhauner committed:
Subscribe even later

paulhauner committed:
Fix slashing test compile errors

paulhauner committed:
Fix slashing db compile error

paulhauner committed:
Add comments, fancy block retry logic

paulhauner committed:
Fix test compile error

paulhauner committed:
Tidy

paulhauner committed:
Break proposer duties into own file, refactor

paulhauner committed:
Avoid bouncing pubkey lock

paulhauner committed:
Improve test coverage, fix bugs

paulhauner committed:
Test for proposer cache changes

paulhauner committed:
Test proposer duties from future epoch

paulhauner committed:
Add first pass of attn duties refactor

paulhauner committed:
Fix failing test

paulhauner committed:
Cache next-epoch duties during state advance

paulhauner committed:
Update comment

paulhauner committed:
Add more debugging for shuffling cache

paulhauner committed:
Reduce logging for cache

paulhauner committed:
Add more VC metrics

paulhauner committed:
Avoid state read for committee cache

paulhauner committed:
More VC metrics

paulhauner committed:
Fix clippy lint

paulhauner committed:
Misc tidying

paulhauner committed:
Merge branch 'unstable' into faster-duties

paulhauner committed:
Fix compile failure

paulhauner committed:
Add first pass at state advance functions

paulhauner committed:
Start threading state advance into beacon chain

paulhauner committed:
Replace per_slot_processing in test_utils

paulhauner committed:
Update http_api for state_advance

paulhauner committed:
Thread harness changes through tests

paulhauner committed:
Add docs, fix clippy lints

paulhauner committed:
Undo changes to store

paulhauner committed:
Add warning comments about inconsistent state

paulhauner committed:
Tidy http_api

paulhauner committed:
Unify decision root logic

paulhauner committed:
Add log in duties service

paulhauner commented:
This is finally ready for review :tada: I'm going to start running it on Pyrmont now (previous iterations of this PR have been run there before).

michaelsproul commented on beacon_node/network/src/beacon_processor/tests.rs:
```diff
@@ -92,15 +92,16 @@ impl TestRig {
                 harness.chain.slot().unwrap(),
             )
             .into_iter()
-            // .map(|vec| vec.into_iter().map(|(attestation, _subnet_id)| attestation))
```
Commented out boi

michaelsproul commented on validator_client/src/attestation_service.rs:
```diff
@@ -355,52 +352,36 @@ impl<T: SlotClock + 'static, E: EthSpec> AttestationService<T, E> {
 
             // Ensure that the attestation matches the duties.
-            if duty_slot != attestation_data.slot || duty_committee_index != attestation_data.index
+            #[allow(clippy::suspicious_operation_groupings)]
```
If this lint crops up a lot I think we should nuke it, how silly!

michaelsproul commented on beacon_node/http_api/src/proposer_duties.rs:
```diff
@@ -0,0 +1,281 @@
+        .map_err(warp_utils::reject::beacon_chain_error)?;
+
+    // We can supply the genesis block root as the block root since we know that the only block that
+    // decides its own root is the genesis block.
```
I had to think about this for a bit, and I'm considering ways to improve the wording.. maybe: _the only block which decides its own shuffling is the genesis block_? It's kind of unlucky we have to deal with this at all here, but I can't see a way of pushing this concern down to the `BeaconState`.

michaelsproul commented on beacon_node/http_api/src/proposer_duties.rs:
```diff
@@ -0,0 +1,281 @@
+        Err(warp_utils::reject::custom_server_error(format!(
+            "{} proposers is not enough for {} slots",
+            slots_per_epoch,
+            proposer_data.len()
```
these args should be round the other way

michaelsproul commented on validator_client/src/duties_service.rs:
```diff
@@ -1,758 +1,755 @@
+            loop {
+                match duties_service.slot_clock.duration_to_next_slot() {
+                    Some(duration) => {
+                        // Run this poll before the wait, this should hopefully download all the indices
+                        // before the block/attestation tasks need them.
+                        poll_validator_indices(&duties_service).await;
```
I don't know if it's just because I've been thinking about clock lag too much recently, but I wonder if we risk misalignment issues by getting the duration, doing some work, then waiting the duration? What if we put the poll before the `match` instead?

paulhauner replied:
Good call!

michaelsproul commented on validator_client/src/duties_service.rs:
```diff
@@ -1,758 +1,755 @@
+    let mut already_warned = Some(());
+    let mut attesters_map = duties_service.attesters.write();
+    for duty in relevant_duties {
+        let proposer_map = attesters_map.entry(duty.pubkey).or_default();
```
Typo: proposer_map -> attester_map

paulhauner replied:
Copy-paste exposed :flushed:

michaelsproul commented on validator_client/src/duties_service.rs:
```diff
@@ -1,758 +1,755 @@
+/// However, we also have the slashing protection as a second line of defence. These two factors
+/// provide an acceptable level of safety.
+///
+/// It's important to note that since there is a 0-epoch look-ahead (i.e., no look-ahead) for block
+/// proposers then it's very likely that a proposal for the first slot of the epoch will need go
+/// through the slow path every time. I.e., the proposal will only happen after we've been able to
+/// download and process the duties from the BN. This means it is very important to ensure this
+/// function is as fast as possible.
```
I agree with everything written here, but I wonder if at some point in the future it might be worth experimenting with polling partway through the end of the previous slot (i.e. once the dependent-root block has arrived). Or is this obsoleted by some future API on the BN that pushes re-org info to the VC?

paulhauner replied:
Yeah, I totally agree that's worth experimenting with in the future. I don't know of a standard API event that would work well here, since we make no guarantees about the proposer shuffling being known for the next epoch after we get a block. I think the best we could do is add our own event (not super keen on this, tbh) or just make an assumption about when the BN should already know then shuffling and then poll then.

One of the benefits of checking the proposer shuffling early would be that we prime the caches for the scenario where the last block of the epoch is missed.

nflaig replied:
Hey, wondering if there has been further discussions on this. Similar logic is also implemented in Lodestar but facing some issues due to that https://github.com/ChainSafe/lodestar/issues/5314. Skipping a slot seems to cause more harm then good in a lot of cases. Lodestar applies this logic also for other routines such as attestations and sync committee

michaelsproul reviewed (approved):
Beautiful all around, very few nitpicks from me this time and just a few musings/ramblings instead :stuck_out_tongue: 

I think this solves #2052 only for the duties service, meaning that #2162 will still be required (although whichever merges second will need to resolve conflicts).

michaelsproul commented:
Addendum: I had a look for cache misses on one of the Pyrmont nodes running this code and found:

* All 3 cache misses I checked were due to missed blocks at the end of the previous epoch causing the timed state advance to fail (thus not priming the cache).
* Curiously, in this case it's the `StateAlreadyAdvanced` log that trips: because the state from slot 30 (mod 32) has been advanced to slot 31 and is sitting in the snapshot cache, but will not be advanced further. This seems fine, I just found it a bit surprising.

paulhauner committed:
Address comments by @michaelsproul

paulhauner committed:
Merge branch 'pr/2162' into faster-duties

paulhauner committed:
Poll fork service before wait

paulhauner commented:
As discussed @michaelsproul, I've addressed your comments and merged #2162 into this branch.

You can see the #2162 specific changes here: https://github.com/sigp/lighthouse/compare/aea2ffb...4824afd

michaelsproul commented on validator_client/src/notifier.rs:
```diff
@@ -11,86 +11,88 @@ pub fn spawn_notifier<T: EthSpec>(client: &ProductionValidatorClient<T>) -> Resu
-                    "synced" => num_synced,
-                )
-            }
+            match duties_service.slot_clock.duration_to_next_slot() {
```
We could fight a bit of rightward drift by using an `if let` here (it's already pretty nested). But happy to merge either way.

paulhauner committed:
Remove some match nesting

paulhauner commented:
I cancelled CI, bors can do it.

So excited to get this merged, thanks for the review!!

bors r+

paulhauner commented:
I'll batch this

bors r-

paulhauner commented:
bors r+
